### ISSUE TITLE
Issues getting this to work

### ISSUE THREAD

**Issue by kennex-software** [ushills/msgtopdf#4]
Hi! From Reddit:
- I am unable to even import Msgtopdf for some reason. I cannot import it in a .py file or with cmd. It gives me an import error.
- if I just use the cmd to try to convert .msg files, it tells me everything: "Filename is invalid, enter a valid filename!", even though I put the correct path and have it in quotes (which I assume is necessary, but it doesn't state to anywhere). I tried this with -d and -f, neither work. When I do this, it does create a folder with the images, attachments, and html file, but does not convert anything to a PDF.
- I have Outlook installed.
- I have wkhtmltopdf installed and in PATH. Version 0.12.5

1. I have installed via "pip install msgtopdf" in cmd.
2. wkhtmltopdf in cmd:
![image](https://user-images.githubusercontent.com/53454116/84436980-512c6f00-ac02-11ea-8a51-2c27ee00d5be.png)
3. msg2pdf in cmd:
![image](https://user-images.githubusercontent.com/53454116/84437056-66a19900-ac02-11ea-9a0e-08c3a094ac3d.png)
4. I do currently have Python 3.7.7 installed.  This was installed from Python.org.

Most import errors I've come across seem to be due to circle references, but I could not find any such references in your code.  That's not to say they don't exist. 

Edit:
This is the error I get in console. 
![image](https://user-images.githubusercontent.com/53454116/84437537-2858a980-ac03-11ea-879d-79e8387d339e.png)

**Comment by ushills** [comment]
Would you be able to try the example in tests/example.py, but substitute and change the email filename as appropriate:

`from msgtopdf import Msgtopdf`

with

`from msgtopdf.msgtopdf import Msgtopdf`

I have fixed that locally and will improve the error messages if I can see
what the problem is your end. It seems to be a problem with wkhtmltopdf..

Which Outlook are you using?

**Comment by kennex-software** [comment]
Changing to `from msgtopdf.msgtopdf import Msgtopdf` allows me to bypass the error!  That works.  

Now I am getting this error though, running through.  

> "C:\Program Files\Python37\python.exe" "E:/Programming Projects/Emails/messagetopdf.py"
> Traceback (most recent call last):
>   File "E:/Programming Projects/Emails/messagetopdf.py", line 16, in <module>
>     main()
>   File "E:/Programming Projects/Emails/messagetopdf.py", line 12, in main
>     email.email2pdf()
>   File "C:\Program Files\Python37\lib\site-packages\msgtopdf\msgtopdf.py", line 56, in email2pdf
>     PurePath(self.save_path, self.file_name + ".pdf"),
>   File "C:\Program Files\Python37\lib\subprocess.py", line 488, in run
>     with Popen(*popenargs, **kwargs) as process:
>   File "C:\Program Files\Python37\lib\subprocess.py", line 800, in __init__
>     restore_signals, start_new_session)
>   File "C:\Program Files\Python37\lib\subprocess.py", line 1148, in _execute_child
>     args = list2cmdline(args)
>   File "C:\Program Files\Python37\lib\subprocess.py", line 555, in list2cmdline
>     needquote = (" " in arg) or ("\t" in arg) or not arg
> TypeError: argument of type 'PureWindowsPath' is not iterable

If I already have the folder made (from running the first time), I get this error:

> "C:\Program Files\Python37\python.exe" "E:/Programming Projects/Emails/messagetopdf.py"
> Traceback (most recent call last):
>   File "E:/Programming Projects/Emails/messagetopdf.py", line 16, in <module>
>     main()
>   File "E:/Programming Projects/Emails/messagetopdf.py", line 12, in main
>     email.email2pdf()
>   File "C:\Program Files\Python37\lib\site-packages\msgtopdf\msgtopdf.py", line 32, in email2pdf
>     Path.mkdir(Path(self.save_path))
>   File "C:\Program Files\Python37\lib\pathlib.py", line 1268, in mkdir
>     self._accessor.mkdir(self, mode)
> FileExistsError: [WinError 183] Cannot create a file when that file already exists: 'E:\\Programming Projects\\Emails\\msg2'

I can easily delete the folder with the HTML, so that's not a huge deal.

**Comment by ushills** [comment]
If you delete the created folder and run the command msg2pdf or use the
example script with the previous correction does it create the HTML file
and images or the full PDF?

If it creates the HTML files then there is an issue with calling
wkhtmltopdf within the module.

I noticed you are using another drive path i.e E, is that a Windows drive?
If so I need to check if Pathlib works across file boundaries.  Can you try
running it again with the module and msg file on the same drive?

**Comment by kennex-software** [comment]
I tried to run this one C:

> "C:\Program Files\Python37\python.exe" C:/Users/patri/emailscript/messagetopdf.py
> Traceback (most recent call last):
>   File "C:/Users/patri/emailscript/messagetopdf.py", line 16, in <module>
>     main()
>   File "C:/Users/patri/emailscript/messagetopdf.py", line 12, in main
>     email.email2pdf()
>   File "C:\Program Files\Python37\lib\site-packages\msgtopdf\msgtopdf.py", line 56, in email2pdf
>     PurePath(self.save_path, self.file_name + ".pdf"),
>   File "C:\Program Files\Python37\lib\subprocess.py", line 488, in run
>     with Popen(*popenargs, **kwargs) as process:
>   File "C:\Program Files\Python37\lib\subprocess.py", line 800, in __init__
>     restore_signals, start_new_session)
>   File "C:\Program Files\Python37\lib\subprocess.py", line 1148, in _execute_child
>     args = list2cmdline(args)
>   File "C:\Program Files\Python37\lib\subprocess.py", line 555, in list2cmdline
>     needquote = (" " in arg) or ("\t" in arg) or not arg
> TypeError: argument of type 'PureWindowsPath' is not iterable

It also made a folder with an HTML file.  It seems like I am unable to get the PDF.  I'm pretty sure that's the issue.  The HTML is working fine.

**Comment by ushills** [comment]
Thank you for the update.

From the log it is definitely an issue with the subprocess.run used to call
wkhtmltopdf.

This may be due to an update with Python. I have a number of Python 3.8
projects using subprocess.call and subprocess.run successfully so it is
something I can solve.

I will review the code and provide an update.

As I currently do not have a working Outlook version would you be willing
to test a new branch on your machine before I update it on PyPi?

>

**Comment by ushills** [comment]
I have issued v0.1.2 to PyPi and tested it on a clean Windows 10 development version and it works here. 
I have slightly improved the error message, but there is more to do on this. 
Can you update to the latest version and test it with your files.  The example.py file also works now, but note you will need to copy it into a folder away from the git clone version if you also have this on your machine. 
Let me know and I can close the issue.

**Comment by kennex-software** [comment]
Hi, I uninstalled the old version of msgtopdf with pip and reinstalled.  All went successfully.  I just tried it with the following code:

`from msgtopdf import Msgtopdf
from pathlib import Path

def main():
    directory = Path.cwd()
    msgfile = "msg2.msg"
    msgfile = Path(directory, msgfile)
    email = Msgtopdf(msgfile)
    email.email2pdf()

if __name__ == "__main__":
    main()`

Where msg2.msg is the file I have in the same directory.  Now when I run this, only the folder is created, no HTML files, and I get this error:

> 06/15/2020 07:46:14 AM - CRITICAL - Could not call wkhtmltopdf

I tried to upgrade to the latest wkhtmltopdf:

![image](https://user-images.githubusercontent.com/53454116/84654835-da32f700-aedd-11ea-8a55-73599884b4b8.png)

But the same error occurs.  Msg2.msg is plain text with no attachments.  I also tested the wkhtmltopdf script in cmd and it works fine.  But when I try the msg2pdf script in cmd, I get this:

![image](https://user-images.githubusercontent.com/53454116/84654489-43663a80-aedd-11ea-95a1-c5da7deaf081.png)

**Comment by ushills** [comment]
I have updated the version on PyPi, would you mind updating to the latest version?
I have added a test to check that wkhtmltopdf is correctly in the path and can be found by Python, as the path being incorrect is the only way I can replicate the error.
Please let me know what this returns as it should now provide a more specific CRITICAL error and output the full path.
Send me any errors and I will take a look.

**Comment by kennex-software** [comment]
I uninstalled and reinstalled via pip.

When I try this in CMD I get this:
![image](https://user-images.githubusercontent.com/53454116/86181215-9c4bea80-bafb-11ea-872f-c8759de8c548.png)

This creates the msg2 folder, but has nothing inside, no HTML or PDF.  This msg2.msg file is plain-text. 

Trying this in a .py file pulled the same error:
![image](https://user-images.githubusercontent.com/53454116/86181483-13817e80-bafc-11ea-90a3-7367a1b9b88b.png)

It also creates the folder, but has nothing inside. 

Let me know if this helps.

**Comment by ushills** [comment]
Please try copying and running the following gist as a python file on your machine and with your environment.

https://gist.github.com/ushills/6318e4d53afed99781cc237075ebc5dd

The only issue seem to be the subprocess.run call to the wkhtmltopdf does not run successfully for you.

This uses the defaults in my module by downloads the PSF code of conduct and converts it to a pdf.

This should help debug where the problem lies, the only other issue I can think of is an invalid piece of HTML in the extracted email that wkhtmltopdf cannot convert but let's try this first.

Let me know what the outcome is.

**Comment by kennex-software** [comment]
Hi, sorry for the delay.  I just did this test and this is the result:

> CompletedProcess(args=['wkhtmltopdf', '--log-level', 'info', '--encoding', 'utf-8', '--footer-font-size', '6', '--footer-line', '--footer-center', '[page] / [topage]', 'https://www.python.org/psf/conduct/', 'python_cod.pdf'], returncode=0, stdout=b'', stderr=b'Loading pages (1/6)\r\n[>                                                           ] 0%\r[======>                                                     ] 10%\r[========>                                                   ] 14%\r[==========>                                                 ] 17%\r[===========>                                                ] 19%\r[=============>                                              ] 22%\r[==============>                                             ] 24%\r[================>                                           ] 27%\r[=================>                                          ] 29%\r[============================>                               ] 48%\r[==================================>                         ] 58%\r[========================================>                   ] 67%\r[==============================================>             ] 77%\r[===============================================>            ] 79%\r[=================================================>          ] 82%\r[===================================================>        ] 85%\r[====================================================>       ] 88%\r[======================================================>     ] 90%\r[======================================================>     ] 90%\r[======================================================>     ] 90%\r[============================================================] 100%\rCounting pages (2/6)                                               \r\n[============================================================] Object 1 of 1\rResolving links (4/6)                                                       \r\n[============================================================] Object 1 of 1\rLoading headers and footers (5/6)                                           \r\nPrinting pages (6/6)\r\n[>                                                           ] Preparing\r[=======>                                                    ] Page 1 of 8\r[===============>                                            ] Page 2 of 8\r[======================>                                     ] Page 3 of 8\r[==============================>                             ] Page 4 of 8\r[=====================================>                      ] Page 5 of 8\r[=============================================>              ] Page 6 of 8\r[====================================================>       ] Page 7 of 8\r[============================================================] Page 8 of 8\rDone                                                                      \r\n')

So that seems to have worked well.  It did create a PDF called python_cod.pdf and created an 8 page PDF with images and everything.  But if I try the code, it's still having issues calling the wkhtmltopdf.

**Comment by ushills** [comment]
Well that indicates that python can correctly call wkhtmltopdf and the options work. 

The only way I can see working through this is running each method in turn and seeing where it fails, I could be a corruption in the created html that prevents wkhtmltopdf producing the pdf. 

I will set up another gist and send that to you over the weekend.

**Comment by ushills** [comment]
Thanks for your patience.

I have created a gist to manually check each step to see where the problem lies.

Can you copy and paste each of the commands in the linked Gist into a python prompt and let me know any steps that fail or if the outputs at any stage are incorrect.

This will really help me with debugging what is going wrong.

https://gist.github.com/ushills/6c2910fec7a01c866d033395021bc635

**Comment by kennex-software** [comment]
I copied this file directly and got this error:

```
"C:\Program Files\Python37\python.exe" "E:/Programming Projects/Emails/manual_check.py"
Traceback (most recent call last):
  File "E:/Programming Projects/Emails/manual_check.py", line 32, in <module>
    subprocess.run(["wkhtmltopdf", "--log-level", "warn", "--encoding", "utf-8", "--footer-font-size", "6", "--footer-line", "--footer-center", "[page] / [topage]", email.html_body_file, PurePath(email.save_path, email.file_name + ".pdf"),])
  File "C:\Program Files\Python37\lib\subprocess.py", line 488, in run
    with Popen(*popenargs, **kwargs) as process:
  File "C:\Program Files\Python37\lib\subprocess.py", line 800, in __init__
    restore_signals, start_new_session)
  File "C:\Program Files\Python37\lib\subprocess.py", line 1148, in _execute_child
    args = list2cmdline(args)
  File "C:\Program Files\Python37\lib\subprocess.py", line 555, in list2cmdline
    needquote = (" " in arg) or ("\t" in arg) or not arg
TypeError: argument of type 'PureWindowsPath' is not iterable

Process finished with exit code 1
```

I had not looked into it.  The HTML file DID get created.  It did not create a PDF.  Let me know if this helps!

**Comment by ushills** [comment]
Please try version v.0.1.5 updated on PyPi.

This should now have fixed the problem as your report was really helpful.

If if fails can you try the Gist again?

**Comment by kennex-software** [comment]
We have success!  With text!  Woohoo! 

But unfortunately, issues with images. 

I tried the normal program:
```
from msgtopdf.msgtopdf import Msgtopdf
from pathlib import Path


def main():
    directory = Path.cwd()
    msgfile = "emailmessage.msg"
    msgfile1 = Path(directory, msgfile)
    email = Msgtopdf(msgfile1)
    email.email2pdf()


if __name__ == "__main__":
    main()
```

And this printed it to PDF as was supposed to happen, but images (are not saved) are shown in the PDF like this:
![image](https://user-images.githubusercontent.com/53454116/88479377-c6d36c80-cf1c-11ea-9d59-1f3647df8117.png)

I also got this error:
```
"C:\Program Files\Python37\python.exe" "E:/Programming Projects/Emails/messagetopdf.py"
Warning: Blocked access to file E:/Programming Projects/Emails/emailmessage/image001.jpg
Warning: Blocked access to file E:/Programming Projects/Emails/emailmessage/image002.jpg
Warning: Blocked access to file E:/Programming Projects/Emails/emailmessage/image005.jpg
Warning: Blocked access to file E:/Programming Projects/Emails/emailmessage/image006.jpg
Error: Failed to load about:blank, with network status code 301 and http status code 0 - Protocol "about" is unknown
Error: Failed to load about:blank, with network status code 301 and http status code 0 - Protocol "about" is unknown
Error: Failed to load about:blank, with network status code 301 and http status code 0 - Protocol "about" is unknown
Error: Failed to load about:blank, with network status code 301 and http status code 0 - Protocol "about" is unknown
Exit with code 1 due to network error: ProtocolUnknownError

Process finished with exit code 0
```

If I run this with the Gist, all of the images are downloaded (they are all .jpg) and the HTML file is created, but I get this error:

```
"C:\Program Files\Python37\python.exe" "E:/Programming Projects/Emails/manual_check.py"
Traceback (most recent call last):
  File "E:/Programming Projects/Emails/manual_check.py", line 29, in <module>
    subprocess.run(["wkhtmltopdf", "--log-level", "warn", "--encoding", "utf-8", "--footer-font-size", "6", "--footer-line", "--footer-center", "[page] / [topage]", email.html_body_file, PurePath(email.save_path, email.file_name + ".pdf"),])
  File "C:\Program Files\Python37\lib\subprocess.py", line 488, in run
    with Popen(*popenargs, **kwargs) as process:
  File "C:\Program Files\Python37\lib\subprocess.py", line 800, in __init__
    restore_signals, start_new_session)
  File "C:\Program Files\Python37\lib\subprocess.py", line 1148, in _execute_child
    args = list2cmdline(args)
  File "C:\Program Files\Python37\lib\subprocess.py", line 555, in list2cmdline
    needquote = (" " in arg) or ("\t" in arg) or not arg
TypeError: argument of type 'PureWindowsPath' is not iterable

Process finished with exit code 1
```

It's super helpful for just the text though!  As that works...

**Comment by ushills** [comment]
Can you try adding the following to the subprocess call.

--enable-local-file-access


So that the full command is.

subprocess.run(["wkhtmltopdf", "--log-level", "warn", "--encoding",
"utf-8", "--enable-local-file-access", "--footer-font-size", "6",
"--footer-line", "--footer-center", "[page] / [topage]",
email.html_body_file, str(PurePath(email.save_path),
str(email.file_name + ".pdf")),])

**Comment by ushills** [comment]
I have released v0.1.6 that addresses the local file access issue, can you please try this version this should now fix all the issues you have been experiencing.

Thank you for your help, I would not have been able to debug this locally as I could not replicate the issues on my machine but your debugging showed me what the issues were.

**Comment by kennex-software** [comment]
No!  Thank you!!  Everything works now on my end, perfectly, using this code:

```
from msgtopdf import Msgtopdf
from pathlib import Path

def main():
    directory = Path.cwd()
    msgfile = "emailmessage.msg"
    msgfile1 = Path(directory, msgfile)
    email = Msgtopdf(msgfile1)
    email.email2pdf()

if __name__ == "__main__":
    main()
```

Closing this!  Thank you so much!
